src/routing/middleware/cors.rs

```diff
@@ -1,80 +1,544 @@
+    // with default headers set.
+    if method == Method::OPTIONS && !is_valid_preflight_response(&response) {
+        state.invalid_canisters.insert(canister_id, ());
+        return state.default_preflight_response();
+    }
+
+    state.apply_cors(method, &mut response);
+    response
+}
+
+pub fn layer(max_age: Duration, allow_origin: Vec<HeaderValue>) -> CorsLayer {
+    let mut layer = CorsLayer::new()
+        .expose_headers(EXPOSE_HEADERS)
+        .allow_headers(ALLOW_HEADERS)
+        .max_age(max_age);
+
+    if allow_origin.len() == 1 && allow_origin[0] == "*" {
+        layer = layer.allow_origin(Any);
+    } else {
+        layer = layer.allow_origin(allow_origin);
+    }
+
+    layer
 }
 
-pub fn http_gw_layer(methods: &[Method]) -> CorsLayer {
-    CorsLayer::new()
-        .allow_origin(Any)
-        .allow_methods(methods.to_vec())
-        .expose_headers([
-            ACCEPT_RANGES,
-            CONTENT_LENGTH,
-            CONTENT_RANGE,
-            X_REQUEST_ID,
-            X_IC_CANISTER_ID,
-        ])
-        .allow_headers([
-            USER_AGENT,
-            DNT,
-            IF_NONE_MATCH,
-            IF_MODIFIED_SINCE,
-            CACHE_CONTROL,
-            CONTENT_TYPE,
-            RANGE,
-            COOKIE,
-            X_REQUESTED_WITH,
-            X_IC_CANISTER_ID,
-            X_OC_JWT,
-            X_OC_API_KEY,
-        ])
-        .max_age(10 * MINUTE)
+#[cfg(test)]
+mod test {
+    use crate::principal;
+
+    use super::*;
+    use axum::{Router, body::Body, middleware::from_fn_with_state};
+    use bytes::Bytes;
+    use candid::Principal;
+    use http::StatusCode;
+    use tower::ServiceExt;
+
+    #[tokio::test]
+    async fn test_cors_http() {
+        let s = Arc::new(
+            CorsStateHttp::new(
+                10,
+                Duration::from_secs(600),
+                vec![
+                    HeaderValue::from_static("foo"),
+                    HeaderValue::from_static("bar"),
+                ],
+                Duration::from_secs(7200),
+            )
+            .unwrap(),
+        );
+
+        assert_eq!(s.allow_methods, "HEAD, GET, POST, PUT, DELETE, PATCH");
+        assert_eq!(
+            s.allow_headers,
+            "user-agent, dnt, if-none-match, if-modified-since, cache-control, content-type, range, cookie, x-requested-with, x-ic-canister-id, x-oc-jwt, x-oc-api-key"
+        );
+        assert_eq!(s.allow_origin, "foo, bar");
+        assert_eq!(
+            s.expose_headers,
+            "accept-ranges, content-length, content-range, x-request-id, x-ic-canister-id"
+        );
+        assert_eq!(s.max_age, "7200");
+        assert_eq!(
+            s.vary,
+            "origin, access-control-request-method, access-control-request-headers"
+        );
+
+        let canister_id = CanisterId(principal!("aaaaa-aa"));
+
+        // Check that the existing response headers are not overriden
```

Comment:
```suggestion
        // Check that the existing response headers are not overridden
```
But I would move this down to line 328 as here you are just setting up the router that will set some headers on the responses and only later you check that the headers are not overridden.

Reply:
Actually, now reading on, I see what you do: first check for pre-flight and normal request that headers are not overridden and then check that they are overridden later. So, let's just leave it.